### DISCUSSION TITLE
Help with multi-hierarchy pattern

### DISCUSSION THREAD

**Discussion by lukuser** [Budibase/budibase#9735] [Help]
Hello,
I'm working on a deal specification sheet in a brokerage environment.  Each deal has many deal members and each deal member can participate on many deals. This many-to-many relationship is represented through dealpart table with additionnal fields, such as commission split, gross commission, member category, etc. Each deal member is stored in user table and can have a company record assigned to it. The structure is like so:

deal  (one -to-many) dealparty (many-to-one) user (many-to-one) company (many-to-one) address

The form for the deal/:id is a multi step form where defining the deal members  is one of the steps (there are other aspect to the deal such as invoices , terms, etc in other steps).

I'm going in circles trying to implement the pattern of adding  a new deal member or editing one of the existing ones. Currently in the step there's a table connected to a query showing existing members (the user, it's company with address, and some dealparty props). Clicking on a row would edit the member record or the button would add a new member. 

In edit mode, I need to be able to change the user of the dealparty and/or edit dealparty properties. For the user, I need to pick from all existing users or add a new one. If adding a new one, I need to pick from existing set of companies or add a new company. For simplicity, I assume that the address of the company will always be new. 
Similarly, in New Member mode,  I need to create a new dealpartyrecord. For that record, as in the edit mode, I need to pick from all existing users or add a new one. If adding a new one, I need to pick from existing set of companies or add a new one. 

I tried starting with side panels.  I was expecting that each panel would set a state variable with the choice of company or user and, upon closing, would refresh the calling panel with that choice, where the operator would complete the entry of other fields, and save, which would return all the way to dealparty side panel, and saving there would finally update the table on the parent stepform.
In the side panels for the dealparty (new or edit) I had a button to save the record or to select a member. "Select a member" would open a new side panel, call it UserPickerSidePanel. In that panel I had list of existing users (cards under repeater with data source set to query of users joined with company joined with address and some other lookup tables) and a button to add new user. This button would launch another side panel NewUserSidePanel with Create Form set to User table and a button to pick existing Company. That would open another side panel CompanyPickerSidePanel showing  In that panel I had list of existing companies (cards under repeater with data source set to query of companies joined with address and some other lookup tables) and a button to add new company. Finally that button would open another side panel to add a new company NewCompanySidePanel. Unfortunately this didn't work out as closing the last panel would close them all and any entry would be lost. 

I was thinking of doing modal forms, but the issue there is that even if I'd be able to chain the forms just like the panels, I'd not be able to refresh the dealparty data provider of the table on my parent form, since the modal form doesn't trigger refresh of a query-based data sources on the parent form - issue described in https://github.com/Budibase/budibase/issues/9524 which I submitted for a similar problem in another form step (In that case, the solution was to use a side panel since I only needed one level of side panels as I was adding a company).

Finally I was considering using regular forms and navigate to the parent form, but there the issue is that I need to go back to step 6 of the parent form and not start back on step one. I didn't find any way to navigate to a specific form step.  I'm running out of ideas.. Did anyone encounter a similar pattern? Is chaining side panels like that doable or am I dreaming in color?

![image](https://user-images.githubusercontent.com/81827963/219877196-5e57287d-69dc-4a6a-ab46-08379e3a2e37.png)

**Comment by melohagan** [comment]
Hey @lukuser 

For saving forms with more complex relationship setups, creating a screen with a form will probably work better than using side panels.

There's a couple of related tutorials here:
 - https://docs.budibase.com/docs/form-steps
 - https://docs.budibase.com/docs/chaining-actions#example-creating-and-linking-two-table-entries-in-one-form
 - https://github.com/Budibase/budibase/discussions/5299#discussioncomment-2530261

As for the question regarding navigation to a specific step, currently you cannot bind a value to the *Go to step* input. There's an open issue here: https://github.com/Budibase/budibase/issues/9323

You can see from my second comment in that issue that there is a workaround using conditional UI.
